### DISCUSSION TITLE
Expand JSON value to table

### DISCUSSION THREAD

**Discussion by polyrand** [duckdb/duckdb#7873] [Q&A] [answered]
One thing I love about DuckDB is that it expands JSONL (newline-delimited JSON) to a table with multiple columns. I was wondering if there's a way to do this with arbitrary JSON objects. My main objective is storing JSON objects in a SQLite databases, but then reading them using DuckDB. But I'd like to read them as a table with multiple columns, instead of as a table with a single column containing the full JSON object.

Example code:

```sh
#!/usr/bin/env bash

set -euo pipefail

rm -f test.sqlite3*
rm -f test.jsonl

# Load data to JSONL

echo '{"foo": "bar", "hello": "world"}' >> test.jsonl
echo '{"foo": "baz", "hello": "world"}' >> test.jsonl
echo '{"foo": "qux", "hello": "world"}' >> test.jsonl

# Load data to SQLite

sqlite3 test.sqlite3 <<EOF
DROP TABLE IF EXISTS test;

CREATE TABLE test (
    id INTEGER PRIMARY KEY,
    data TEXT
);

-- insert JSON objects
INSERT INTO test (data) VALUES ('{"foo": "bar", "hello": "world"}');
INSERT INTO test (data) VALUES ('{"foo": "baz", "hello": "world"}');
INSERT INTO test (data) VALUES ('{"foo": "qux", "hello": "world"}');
EOF

# Query JSONL
duckdb <<EOF
SELECT * FROM read_ndjson_auto('test.jsonl');
EOF


# Query SQLite
duckdb <<EOF
INSTALL sqlite_scanner;
LOAD sqlite_scanner;

SELECT data FROM sqlite_scan('test.sqlite3', 'test');

EOF


rm -f test.sqlite3*
rm -f test.jsonl
```

I would like the output of `SELECT data FROM sqlite_scan('test.sqlite3', 'test');` to be the same as `SELECT * FROM read_ndjson_auto('test.jsonl');` (a table).

Maybe I can hack something around using `json_keys` + `json_extract`. But it would still be less "correct" than what you get from `read_ndjson_auto`.

Is this possible?

**Comment by polyrand** [comment]
I was also trying to convert the JSON to a struct, with something like:

```sql
SELECT json(data)::STRUCT(json_extructure(data)) FROM sqlite_scan('test.sqlite3', 'test');
```

And then using `UNNEST` if I manage to convert the JSON object to a struct. But I still haven't figured out how to do this.

I can make it work by explicitly converting to a `STRUCT`. But I'm looking for a way to auto-detect the `STRUCT` types instead of having to set them manually:

```sql
SELECT UNNEST(json(data)::STRUCT(foo TEXT, hello TEXT)) FROM sqlite_scan('test.sqlite3', 'test');
```

i.e: I need to find a way to auto-infer `foo TEXT, hello TEXT`

**Comment by lnkuiper** [comment, marked as answer]
@polyrand Would this work for you?
```sql
CREATE TABLE test (
    data TEXT
);
INSERT INTO test (data) VALUES ('{"foo": "bar", "hello": "world"}');
INSERT INTO test (data) VALUES ('{"foo": "baz", "hello": "world"}');
INSERT INTO test (data) VALUES ('{"foo": "qux", "hello": "world"}');
SELECT transformed.*
FROM (
    SELECT json_transform(data, '{"foo":"VARCHAR","hello":"VARCHAR"}') AS transformed
    FROM test
);
┌─────────┬─────────┐
│   foo   │  hello  │
│ varchar │ varchar │
├─────────┼─────────┤
│ bar     │ world   │
│ baz     │ world   │
│ qux     │ world   │
└─────────┴─────────┘
```
We can't do auto-detection in these cases, so you will have to supply the JSON structure as a parameter.

The JSON structure can be obtained like so:
```sql
SELECT json_group_structure(data::JSON) AS structure FROM test;
┌─────────────────────────────────────┐
│              structure              │
│                json                 │
├─────────────────────────────────────┤
│ {"foo":"VARCHAR","hello":"VARCHAR"} │
└─────────────────────────────────────┘
```

**Reply by polyrand** [reply]
Thanks! I think this kind of works. If only I could provide `'{"foo":"VARCHAR","hello":"VARCHAR"}'` as a subselect instead of a constant. But this solves my needs. For reference, in case it's useful for others, this is the Python code I've ended up with:


```python
import duckdb

conn = duckdb.connect()

conn.install_extension("sqlite_scanner")
conn.load_extension("sqlite_scanner")

r = conn.execute(
    """
SELECT transformed.*
FROM (
    SELECT json_transform(data, ?) AS transformed
    FROM sqlite_scan('test.sqlite3', 'test')
)
""",
    (
        conn.query(
            "SELECT json_group_structure(data::JSON) FROM sqlite_scan('test.sqlite3', 'test')"
        ).fetchone()[0],
    ),
)


print(r.fetchall())
```

Or, if you want to use `.query()` and get a `duckdb.DuckDBPyRelation` (for further querying):

```python
json_schema = conn.query(                                                                                                                                      
    "SELECT json_group_structure(data::JSON) FROM sqlite_scan('test.sqlite3', 'test')"                                                         
).fetchone()[0]


conn.query(
    f"""
SELECT transformed.*
FROM (
    SELECT json_transform(data, '{json_schema}') AS transformed
    FROM sqlite_scan('test.sqlite3', 'test')
)
"""
)
```

Thanks again for the help!

**Reply by jylin** [reply]
That works, but why is it not possible to auto-detect columns similar to how read_json is able to do it? Is auto-detection only possible on files?

**Reply by lnkuiper** [reply]
DuckDB is strongly typed, so we need to detect the schema before executing the query. For `read_json` we open the files and do auto-detection while binding. Auto-detecting arbitrary columns is more difficult, because we could have a query like this:
```sql
SELECT * FROM read_json(<subquery>);
```
This would require evaluating the subquery, auto-detecting the types, and then running `read_json` using the auto-detected types. We would have to run all of `<subquery>` while binding the query, before actually executing the full query (including `read_json`), which means that `<subquery>` is executed twice (or once and be cached).

We currently don't have the infrastructure to do this.
